### PR TITLE
fix 6.1 channel passthrough failing

### PR DESCRIPTION
**Pull request by drhill**
6.1 and 4.1 channel passthrough doesn't work on the ShieldTV

### CODE REVIEW

**Comment by drhill** [comment]
This one was only tested on a ShieldTV with a 6.1 channel track.  Both on a receiver connected to a display and just a display.  I believe the movie I tested with was Top Gun.

File: library/src/main/java/com/google/android/exoplayer2/audio/AudioTrack.java [hunk 1]
```diff
@@ -443,6 +443,14 @@ public void configure(String mimeType, int channelCount, int sampleRate,
     }
 
     boolean passthrough = !MimeTypes.AUDIO_RAW.equals(mimeType);
+    if (Util.SDK_INT <= 23 && "foster".equals(Util.DEVICE) && "NVIDIA".equals(Util.MANUFACTURER)) {
+      if (channelConfig == (AudioFormat.CHANNEL_OUT_5POINT1 | AudioFormat.CHANNEL_OUT_BACK_CENTER) )
```
Comment (ojw28) [inline library/src/main/java/com/google/android/exoplayer2/audio/AudioTrack.java]: Any reason not to use `channelCount == 7`, and something like `channelCount >= 3 && channelCount <= 5` for the conditions :)? Seems simpler?

File: library/src/main/java/com/google/android/exoplayer2/audio/AudioTrack.java [hunk 2]
```diff
@@ -443,6 +443,14 @@ public void configure(String mimeType, int channelCount, int sampleRate,
     }
 
     boolean passthrough = !MimeTypes.AUDIO_RAW.equals(mimeType);
+    if (Util.SDK_INT <= 23 && "foster".equals(Util.DEVICE) && "NVIDIA".equals(Util.MANUFACTURER)) {
```
Comment (ojw28) [inline library/src/main/java/com/google/android/exoplayer2/audio/AudioTrack.java:444]: nit - Seems like the new code belongs at L444 as part of the block that's setting `channelConfig`. It's currently sandwiched between the line where `passthrough` is calculated and the line where it's used.

**Review by ojw28 — COMMENTED** [review]
Looks fine aside from nits; thanks.

File: library/src/main/java/com/google/android/exoplayer2/audio/AudioTrack.java (hunk 2 above)
Comment (drhill) [reply library/src/main/java/com/google/android/exoplayer2/audio/AudioTrack.java:444]: changed.  Earlier code used the passthrough variable but that didn't fully fix it.

File: library/src/main/java/com/google/android/exoplayer2/audio/AudioTrack.java (hunk 1 above)
Comment (drhill) [reply library/src/main/java/com/google/android/exoplayer2/audio/AudioTrack.java]: Changed.  Just went with that for explicit intent, but the change is easier to read and as easy to decipher.

File: library/src/main/java/com/google/android/exoplayer2/audio/AudioTrack.java (hunk 2 above)
Comment (andrewlewis) [reply library/src/main/java/com/google/android/exoplayer2/audio/AudioTrack.java:444]: Is this workaround needed for PCM output as well as passthrough? If not, the condition should check if passthrough is true. Is it needed for all passthrough encodings?

**Comment by andrewlewis** [comment]
@drhill Thanks for the contribution! Out of curiosity, what is the failure mode without the workaround? Do some tracks just not get played, or does the AudioTrack throw, or something else?

**Comment by drhill** [comment]
I wrote this a few weeks ago so I'm going off memory (and I'm at work so I can't test), but when it failed the video track refused to load/play.  I don't remember if any error was thrown.  

Edit:  checkAudioTrackInitialized throws an error as the track isn't initialized (well named ;) )

File: library/src/main/java/com/google/android/exoplayer2/audio/AudioTrack.java (hunk 2 above)
Comment (drhill) [reply library/src/main/java/com/google/android/exoplayer2/audio/AudioTrack.java:444]: Yes, it is needed for cases where DTS is decoded to PCM via ffmpeg.

**Comment by ojw28** [comment]
Some further clarification on this would definitely be appreciated; let us know when you're able to try it out again! How were you able to determine that 4 different channel configs were broken, given only one test video ;)? We're pretty keen to get a fix in, but we'd like to make sure it's targeted accurately. Thanks!

**Comment by drhill** [comment]
I'm trying to get some more clips to test with, but I know from the Plex forums that other rare configurations did not work corrections.  I know that 1.0, 2.0, 5.1, and 7.1 work fine and that 6.1 is broken without this.  That leaves 2.1, 3.0, 3.1, 4.0, 4.1, and 5.0 to test.  I'll let you know when I can confirm.

**Comment by drhill** [comment]
I was able to get some 3.0, 4.0, 4.1, 5.0.  An AC3 4.0 track worked fine without this PR so I modified the statement to avoid touching that.  Unfortunately I didn't have a 2.1 and 3.1 track to test with.  I believe a 2.1 track would use this code as it would be listed as 3 channels, whereas with the latest change a 3.1 track wouldn't use the PR code.

In order to work the DTS 5.0, AC3 4.1, DTS 3.0, and DD 3.0 tracks needed the PR.  Without the PR checkAudioTrackInitialized does not get a STATE_INITIALIZED return so we through an InitializationException.

**Comment by drhill** [comment]
Also, this PR is assuming that NVidia fixes this whenever they move to API 24... which may not be the best assumption.

**Comment by ojw28** [comment]
Looking better! Re the earlier point about whether this affects passthrough or whether it also affects standard PCM output:

- I'm assuming your test results above are for the passthrough case where ffmpeg isn't used. Is that correct?
- For the case where you decode to PCM using ffmpeg that you state is also problematic: Are you sure this isn't a more general problem elsewhere? In particular, this could also be caused by `FfmpegDecoder.getChannelCount` returning a channel count that differs from the actual number of channels it produces on the output side. If you could verify whether the issue is Shield specific when using this path, or whether the same issue is present on other devices too, that would remove ambiguity.

Thanks!

**Comment by drhill** [comment]
Actually this happens in passthrough and decode to pcm.

The problem happens when we new our audio track here:

if (sessionId == SESSION_ID_NOT_SET) {
      audioTrack = new android.media.AudioTrack(streamType, sampleRate, channelConfig,
          targetEncoding, bufferSize, MODE_STREAM);
    } else {
      // Re-attach to the same audio session.
      audioTrack = new android.media.AudioTrack(streamType, sampleRate, channelConfig,
          targetEncoding, bufferSize, MODE_STREAM, sessionId);
    }
    checkAudioTrackInitialized();


checkAUdioTrackInitialized throws an error

**Comment by ojw28** [comment]
Could you clarify what error is thrown (a stack trace would be helpful)?

**Comment by drhill** [comment]
12-14 18:38:20.752 8704-9388/com.monsterbutt.homeview E/AudioTrack: Could not get audio output for session 42, stream type -1, usage 1, sample rate 48000, format 0xb000000, channel mask 0x13f, flags 0x1
12-14 18:38:20.752 8704-9388/com.monsterbutt.homeview E/AudioTrack-JNI: Error -22 initializing AudioTrack
12-14 18:38:20.752 8704-9388/com.monsterbutt.homeview E/android.media.AudioTrack: Error code -20 when initializing AudioTrack.
12-14 18:38:20.753 8704-9388/com.monsterbutt.homeview E/ExoPlayerImplInternal: Renderer error.
                                                                               com.google.android.exoplayer2.ExoPlaybackException
                                                                                   at com.google.android.exoplayer2.audio.MediaCodecAudioRenderer.processOutputBuffer(MediaCodecAudioRenderer.java:339)
                                                                                   at com.google.android.exoplayer2.mediacodec.MediaCodecRenderer.drainOutputBuffer(MediaCodecRenderer.java:891)
                                                                                   at com.google.android.exoplayer2.mediacodec.MediaCodecRenderer.render(MediaCodecRenderer.java:484)
                                                                                   at com.google.android.exoplayer2.ExoPlayerImplInternal.doSomeWork(ExoPlayerImplInternal.java:471)
                                                                                   at com.google.android.exoplayer2.ExoPlayerImplInternal.handleMessage(ExoPlayerImplInternal.java:307)
                                                                                   at android.os.Handler.dispatchMessage(Handler.java:100)
                                                                                   at android.os.Looper.loop(Looper.java:148)
                                                                                   at android.os.HandlerThread.run(HandlerThread.java:61)
                                                                                   at com.google.android.exoplayer2.util.PriorityHandlerThread.run(PriorityHandlerThread.java:40)
                                                                                Caused by: com.google.android.exoplayer2.audio.AudioTrack$InitializationException: AudioTrack init failed: 0, Config(48000, 1276, 49152)
                                                                                   at com.google.android.exoplayer2.audio.AudioTrack.checkAudioTrackInitialized(AudioTrack.java:974)
                                                                                   at com.google.android.exoplayer2.audio.AudioTrack.initialize(AudioTrack.java:521)
                                                                                   at com.google.android.exoplayer2.audio.MediaCodecAudioRenderer.processOutputBuffer(MediaCodecAudioRenderer.java:332)
                                                                                   at com.google.android.exoplayer2.mediacodec.MediaCodecRenderer.drainOutputBuffer(MediaCodecRenderer.java:891) 
                                                                                   at com.google.android.exoplayer2.mediacodec.MediaCodecRenderer.render(MediaCodecRenderer.java:484) 
                                                                                   at com.google.android.exoplayer2.ExoPlayerImplInternal.doSomeWork(ExoPlayerImplInternal.java:471) 
                                                                                   at com.google.android.exoplayer2.ExoPlayerImplInternal.handleMessage(ExoPlayerImplInternal.java:307) 
                                                                                   at android.os.Handler.dispatchMessage(Handler.java:100) 
                                                                                   at android.os.Looper.loop(Looper.java:148) 
                                                                                   at android.os.HandlerThread.run(HandlerThread.java:61) 
                                                                                   at com.google.android.exoplayer2.util.PriorityHandlerThread.run(PriorityHandlerThread.java:40) 
12-14 18:38:20.781 8704-8704/com.monsterbutt.homeview I/RRSwitcherReceiver: Plugged
12-14 18:38:20.939 8704-8800/com.monsterbutt.homeview V/RenderScript: 0x82420000 Launching thread(s), CPUs 4
12-14 18:38:22.009 8704-9388/com.monsterbutt.homeview I/ACodec:  process id - 8704
12-14 18:38:22.011 8704-9506/com.monsterbutt.homeview I/OMXClient: Using client-side OMX mux.
12-14 18:38:22.016 8704-9505/com.monsterbutt.homeview I/MediaCodec: [OMX.Nvidia.h264.decode] setting surface generation to 8912898
12-14 18:38:22.020 8704-9506/com.monsterbutt.homeview I/ACodec: Enable timestamp filtering for Video Decoder
12-14 18:38:22.027 8704-9506/com.monsterbutt.homeview D/SurfaceUtils: set up nativeWindow 0x825fd508 for 1920x1080, color 0x106, rotation 0, usage 0x2b00
12-14 18:38:22.051 8704-9515/com.monsterbutt.homeview I/OMXClient: Using client-side OMX mux.
12-14 18:38:22.058 8704-9515/com.monsterbutt.homeview D/NvOsDebugPrintf: Inside NvxLiteH264DecoderLowLatencyInit
12-14 18:38:22.058 8704-9515/com.monsterbutt.homeview D/NvOsDebugPrintf: NvxLiteH264DecoderLowLatencyInit set DPB and Mjstreaming
12-14 18:38:22.058 8704-9515/com.monsterbutt.homeview D/NvOsDebugPrintf: Inside NvxLiteH265DecoderLowLatencyInit
12-14 18:38:22.058 8704-9515/com.monsterbutt.homeview D/NvOsDebugPrintf: NvxLiteH265DecoderLowLatencyInit set DPB and Mjstreaming
12-14 18:38:22.077 8704-9388/com.monsterbutt.homeview E/AudioTrack: Could not get audio output for session 44, stream type -1, usage 1, sample rate 48000, format 0xb000000, channel mask 0x13f, flags 0x1
12-14 18:38:22.078 8704-9506/com.monsterbutt.homeview D/SurfaceUtils: set up nativeWindow 0x825fd508 for 1920x1088, color 0x106, rotation 0, usage 0x2b00
12-14 18:38:22.078 8704-9388/com.monsterbutt.homeview E/AudioTrack-JNI: Error -22 initializing AudioTrack
12-14 18:38:22.078 8704-9388/com.monsterbutt.homeview E/android.media.AudioTrack: Error code -20 when initializing AudioTrack.
12-14 18:38:22.078 8704-9506/com.monsterbutt.homeview W/ACodec: [OMX.Nvidia.h264.decode] setting nBufferCountActual to 14 failed: -1010
12-14 18:38:22.078 8704-9388/com.monsterbutt.homeview E/ExoPlayerImplInternal: Renderer error.
                                                                               com.google.android.exoplayer2.ExoPlaybackException
                                                                                   at com.google.android.exoplayer2.audio.MediaCodecAudioRenderer.processOutputBuffer(MediaCodecAudioRenderer.java:339)
                                                                                   at com.google.android.exoplayer2.mediacodec.MediaCodecRenderer.drainOutputBuffer(MediaCodecRenderer.java:891)
                                                                                   at com.google.android.exoplayer2.mediacodec.MediaCodecRenderer.render(MediaCodecRenderer.java:484)
                                                                                   at com.google.android.exoplayer2.ExoPlayerImplInternal.doSomeWork(ExoPlayerImplInternal.java:471)
                                                                                   at com.google.android.exoplayer2.ExoPlayerImplInternal.handleMessage(ExoPlayerImplInternal.java:307)
                                                                                   at android.os.Handler.dispatchMessage(Handler.java:100)
                                                                                   at android.os.Looper.loop(Looper.java:148)
                                                                                   at android.os.HandlerThread.run(HandlerThread.java:61)
                                                                                   at com.google.android.exoplayer2.util.PriorityHandlerThread.run(PriorityHandlerThread.java:40)
                                                                                Caused by: com.google.android.exoplayer2.audio.AudioTrack$InitializationException: AudioTrack init failed: 0, Config(48000, 1276, 49152)
                                                                                   at com.google.android.exoplayer2.audio.AudioTrack.checkAudioTrackInitialized(AudioTrack.java:974)
                                                                                   at com.google.android.exoplayer2.audio.AudioTrack.initialize(AudioTrack.java:521)
                                                                                   at com.google.android.exoplayer2.audio.MediaCodecAudioRenderer.processOutputBuffer(MediaCodecAudioRenderer.java:332)
                                                                                   at com.google.android.exoplayer2.mediacodec.MediaCodecRenderer.drainOutputBuffer(MediaCodecRenderer.java:891) 
                                                                                   at com.google.android.exoplayer2.mediacodec.MediaCodecRenderer.render(MediaCodecRenderer.java:484) 
                                                                                   at com.google.android.exoplayer2.ExoPlayerImplInternal.doSomeWork(ExoPlayerImplInternal.java:471) 
                                                                                   at com.google.android.exoplayer2.ExoPlayerImplInternal.handleMessage(ExoPlayerImplInternal.java:307) 
                                                                                   at android.os.Handler.dispatchMessage(Handler.java:100) 
                                                                                   at android.os.Looper.loop(Looper.java:148) 
                                                                                   at android.os.HandlerThread.run(HandlerThread.java:61) 
                                                                                   at com.google.android.exoplayer2.util.PriorityHandlerThread.run(PriorityHandlerThread.java:40) 
12-14 18:38:22.079 8704-9506/com.monsterbutt.homeview W/ACodec: [OMX.Nvidia.h264.decode] setting nBufferCountActual to 13 failed: -1010
12-14 18:38:22.092 8704-9506/com.monsterbutt.homeview D/ACodec: REUSING read fence 70 as write fence in cancelBufferToNativeWindow

**Comment by drhill** [comment]
On dev-v2 branch (no modifications)

**Comment by ojw28** [comment]
Thanks!

**Pull request merged**